data/augment.py

```diff
@@ -0,0 +1,768 @@
+        """Method to submit the final request to create the augmented table
+        
+        Args:
+            target_label (str): Column to be predicted from the input table
+            max_candidates_to_join (int): No of candidates to pick for the join
+            max_features_per_candidate (int): No of features to pick from each candidate
+            
+        Returns:
+            tuple(job_id, error)
+        """
+        flag, err = self.validate_request_status()
+        if not flag:
+            return None, err
+        
+        try:
+            url = f"{self.__connection.endpoint}/augment/{self.id}/match"
+            headers = {"Authorization": f"Bearer {self.__connection.access_token}"}
+            params = {
+                "match_job": {
+                    "max_features_per_result": max_features_per_candidate,
+                    "max_results": max_candidates_to_join,
+                    "target_label": target_label
+                }
+            }
+            resp = requests.post(url, data=json.dumps(params), headers=headers, verify=False)
+            resp.raise_for_status()
+            result = resp.json()
+            self.join_job_id = result["id"]
+            return result["id"], None
+        except Exception as e:
+            return None, str(e)
+    
+    @property
+    def join_status(self):
+        """Returns the status of the join request"""
+        flag, err = self.validate_request_status()
+        if not flag:
+            return "", err
+        
+        if not self.join_job_id:
+            return "", "Join not applied"
+        
+        try:
+            url = f"{self.__connection.endpoint}/augment/{self.id}/match/{self.join_job_id}"
+            headers = {"Authorization": f"Bearer {self.__connection.access_token}"}
+            resp = requests.get(url, headers=headers, verify=False)
+            resp.raise_for_status()
+            result = resp.json()
+            return result["status"], None
+        except Exception as e:
+            return "", str(e)
+
+    @property
+    def output(self):
+        """Returns the dataset id of the newly created augmented table"""
+        flag, err = self.validate_request_status()
+        if not flag:
+            return "", err
+        
+        if not self.join_job_id:
+            return "", "Join not applied"
+        
+        if self.join_status[0] != "DONE":
+            return "", "Join operation not completed yet"
+        
+        try:
+            url = f"{self.__connection.endpoint}/augment/{self.id}/match/{self.join_job_id}"
+            headers = {"Authorization": f"Bearer {self.__connection.access_token}"}
+            resp = requests.get(url, headers=headers, verify=False)
+            resp.raise_for_status()
+            result = resp.json()
+            return result["dataset_id"], None
+        except Exception as e:
+            return "", str(e)
+
+
+class AugmentDataset(CustomData):
```

Comment:
where is the CustomData used?  I guess this is needed since data recipe?

Reply:
@surenH2oai `CustomData` is the base class and we are overriding the method `create_data` in the subclass 'AugmentDataset`. DAI will find out the subclass that derives from `CustomData` which in this case is `AugmentDataset` and it will invoke the `create_data` method to get the updated dataframe with original columns + augmented columns